doc/fvctl.1
.ds PN fvctl

.TH fvctl 1 "July 2010" "FlowVisor" "FlowVisor Manual"

.SH NAME
fvctl \- commandline tool to administer FlowVisor

.SH AUTHOR

Rob Sherwood <rob.sherwood@stanford.edu>

.SH SYNOPSIS
.B fvctl
[\fIoptions\fR] \fIcommand [\fIargs\fR...]

.SH DESCRIPTION
The
.B fvctl
program is a command line tool for configuring, debugging, monitoring, and
administering running FlowVisor instances.  It is able to show the
current state of a FlowVisor, including features, configuration, and
flowspace entries.

Every FlowVisor instance runs an XMLRPC server and accepts API calls
as documented in org.flowvisor.api.FVUserAPI.java.  The tool \fBfvctl\fR
is just a command-line client that invokes these calls.  The expectation
is that more user friendly (read: graphical) interfaces will evolve and
eventually replace \fBfvctl\fR.

.SH OPTIONS


.TP 
\fB--url\fR=\fIURL\fR
Specifes the URL where the FlowVisor's XMLRPC server is listenning.  Default is:
\fIhttps://localhost:8080\fR

.TP 
\fB--user\fR=\fIUSER\fR
Specifes the FlowVisor user name to use when connecting to the XMLRPC server.  Default is: 
\fIroot\fR

.TP
\fB--debug\fR=\fItrue\fR
Turns on additional debugging.  Specifically, print the full stack frame on exception.  Default is: 
\fIfalse\fR

.TP
\fB--passwd-file\fR=\fIfile\fR
Instead of prompting for the password, read it from \fIfile\fR.  Default is to prompt for the password.

.SH COMMANDS

With the \fBfvctl\fR program, the expectation is the FlowVisor is up and running, with its XMLRPC
server listenning at \fIURL\fR.  Currently \fBfvctl\fR requires that the commands be camel-cased, 
e.g., \fIlistFlowSpace\fR, not \fIlistflowspace\fR or \fIlIstFlowSPACE\fR.  To be fixed soon.


.PP
These commands are for creating/managing/deleting slices.  Permissions on
modifying a slice are hierarchical: a user can modify their
own slice, as can the user who created the slice, transitively up the
chain to the root slice.

.TP
\fBlistSlices\fR
    Displays the currently configured slices.

.TP
\fBgetSliceInfo\fR \fIslicename\fR
    Displays the \fIslicename\fR's controller URL, contact information, and who created this slice.

.TP
\fBcreateSlice\fR \fIslicename\fR \fIcontroller_url\fR \fIemail\fR
    Creates a new slice.  Note that the \fIslicename\fR cannot contain
    any of the following characters: !:=[] or new lines.  The controller
    URL is of the form tcp:hostname[:port] so "tcp:controller.myco.com"
    or "tcp:127.0.0.1:12345", with the default port being 6633.  The email
    is used as the administrive contact point if there is a problem with
    the slice.

.TP
\fBchangeSlice\fR \fIslicename\fR \fIkey\fR \fIvalue\fR
    Allow a slice user to change values associated with their slice.
    Currently only "contact_email", "controller_host" and "controller_port"
    are changable.

.TP
\fBdeleteSlice\fR \fIslicename\fR
    Deletes a slice and removed all of the flowspace corresponding to the
    slice.  

.TP
\fBchangePasswd\fR \fIslicename\fR
    Change the password for slice \fIslicename\fR.



.PP
The following commands manipulate the FlowVisor's flowspace.
Each flowspace entry consists of a DPID, a FLOW_MATCH rule, a list
of SLICEACTIONS, an ID, and a PRIORITY.  The format of the DPID and
FLOW_MATCH parameters are specified bellow in the DPID and FLOW_MATCH
sections.  A single flow entry makes the statement "The flowspace
described by FLOW_MATCH on switch DPID is controlled by the slices in
the list SLICEACTIONS".  Flow entries are sorted by PRIORITY (high number
is higher priority) and only the highest priority matched is considered.  Priorities are 
any positive integer in the range [0:2^31].  Additionally, each
flow entry is associated with a unique, rule-position-independent ID, that names
that specific rule.  While the XMLRPC API supports batching of add/remove/changing
flowspace rules for transactional changes, \fBfvctl\fR does not.

.TP
\fBlistFlowSpace\fR
    Lists the FlowVisor's flow-based slice policy rules, i.e., the flowspace.  

.TP
\fBremoveFlowSpace\fR \fIID\fR
    Removes rule with id=\fIID\fR.  
    
.TP
\fBaddFlowSpace\fR \fIDPID\fR \fIPRIORITY\fR \fIFLOW_MATCH\fR \fISLICEACTIONS\fR
    Creates a new rule and returns the new rule's ID.  See below for the format of DPID, FLOW_MATCH, and SLICEACTIONS.
.TP
\fBchangeFlowSpace\fR \fIID\fR \fIDPID\fR \fIPRIORITY\fR \fIFLOW_MATCH\fR \fISLICEACTIONS\fR
    Replaces rule ID with a new rule with the specificed parameters.  See below for the format of DPID, FLOW_MATCH, and SLICEACTIONS.
.PP
The following commands provide information on the devices connected to the FlowVisor (e.g., switches)
and the topology that connects them.

.TP
\fBlistDevices\fR
    Displays the devices (e.g., switches) currently connected to the FlowVisor, by DPID.

.TP
\fBgetDeviceInfo\fR \fIdpid\fR
    Displays information about a connected device, including available ports and remote connection.

.TP
\fBgetLinks\fR
    Displays the links between the devices, as inferred by FlowVisor's topology discovery mechanism.

.TP 
\fBregisterCallback\fR \fIURL\fR \fIcookie\fR
Register a XMLRPC callback for the flowvisor to call when there is
a change in the topology.  Once registered, with each topology change
(with some dampenning), FlowVisor will connect to the server listed in
URL, and make the XMLRPC call to the path of \fIURL\fR with \fIcookie\fR
as a parameter.

.TP 
\fBunregisterCallback\fR 
UnRegisters a previously registered XMLRPC callback.

.PP
These commands are a general purpose hook for manipulating all elements
of the FlowVisor's config.  The FlowVisor's config is internally stored
as a tree of options, delimited by exclaimation marks.  Care must be
taken while using them, as they bypass the sanity checks that more
specific commands use.  For example, it is possible to change a slice's
password via 

.B \fBfvctl setConfig\fR slices!\fISLICENAME\fR!passwd_crypt 057432050a34e81e79bd842fec77dcad

but since there are no sanity checks on the password crypt string, it is
possible to accidentally set the password crypt to an invalid string.
These commands can only be run in supervisor mode, which currently is
only the "root" user.

.TP
\fBgetConfig\fR \fIconfigEntry\fR
    List the entire subtree of the FlowVisor's configuration started
    at node \fIconfigEntry\fR.  Note that a literal period is used to
    denote the whole tree, so `\fBfvctl getConfig\fR \fI.\fR` will list
    the FlowVisor's entire configuration.

.TP
\fBsetConfig\fR \fIconfigEntry\fR \fIvalue\fR
    Set the value of a specific config entry to \fIvalue\fR.  The format
    of the value string must correspond to the type of \fIconfigEntry\fR, e.g., 
    a boolean value must be \fItrue\fR or \fIfalse\fR, an integer must be an
    decimal number, etc.  

.PP
The following commands relay how many OpenFlow messages have been
sent/recv/drop'd by switches and slices.  They are currently returned in a
single block string format, which is suboptimal, but will likely go away
if we get rid of XMLRPC.  Any message type not listed implicitly has a
count of zero (never appeared).  The current output looks something like:

.PP
Switch 00:00:00:00:00:00:00:01's STATS:
.br
---Sent---
.br
classifier-dpid=00:00:00:00:00:00:00:01 :: FEATURES_REPLY=3,HELLO=1
.br
Total :: FEATURES_REPLY=3,HELLO=1
.br
---Recv---
.br
slicer_bob_dpid=00:00:00:00:00:00:00:01 :: FEATURES_REQUEST=1
.br
slicer_alice_dpid=00:00:00:00:00:00:00:01 :: FEATURES_REQUEST=1,PACKET_OUT=1
.br
Total :: FEATURES_REQUEST=2,PACKET_OUT=1
.br
---Drop---
.br
Total ::
.br


.TP
\fBgetSliceStats\fR \fIsliceName\fR
    Returns statistics for \fIsliceName\fR in terms of messages sent, received, and dropped, subdivided by message type.

.TP
\fBgetSwitchStats\fR \fIdpid\fR
    Returns statistics for the switch with dpid \fIdpid\fR in terms of messages sent, received, and dropped, subdivided by message type.

.TP
\fBgetSwitchFlowDB\fR \fIdpid\fR
    Dump's the FlowVisor's view of \fIdpid\fR's flow table.  The FlowVisor
    tracks the sum of the flow_mods and flow_removed messages, so this
    DB should be up to date baring race conditions and switch bugs.
    For hardware switches, this is likely much cheaper to ask the FlowVisor
    for its cached copy than to ask the switch itself. Note that track_flows must
    be enabled for this feature to work, else this command will return an empty list.
    Flow tracking can be enabled with:
        \fBfvctl\fR \fIsetConfig\fR \fB!flowvisor!track_flows\fR \fItrue\fR

.TP
\fBgetSliceRewriteDB\fR \fIsliceName\fR \fIdpid\fR
    Dump how the flowvisor has rewritten the slices flowentries.  This command
    outputs how the original flowMod \fIfm_original\fR was rewritten to a 
    sequence of flowMods: \fIfm_rewrite1,fm_rewrite2\fR.  Similar to getSwitchFlowDB,
    this command returns an empty list if track_flows is disabled.

.PP 
The following commands are used for debugging the XMLRPC server
.TP
\fBping\fR \fImsg\fR
    Simply echo's \fImsg\fR back, verifying that the XMLRPC server is up, functioning, and the authentication is working.


.SH "FLOW SYNTAX"

Some \fBfvctl\fR commands accept an argument that describes a flow or
flows.  Such flow descriptions comprise a series
\fIfield\fB=\fIvalue\fR assignments, separated by commas.

The following field assignments describe how a flow matches a packet.
If any of these assignments is omitted from the flow syntax, the field
is treated as a wildcard; thus, if all of them are omitted, the
resulting flow matches all packets.  The string \fBall\fR or \fBany\fR
is used to specify a flow that matches all packets.

.IP \fBin_port=\fIport_no\fR
Matches physical port \fIport_no\fR.  Switch ports are numbered as
displayed by \fBfvctl\fR getDeviceInfo \fIDPID\fR.

.IP \fBdl_vlan=\fIvlan\fR
Matches IEEE 802.1q virtual LAN tag \fIvlan\fR.  Specify \fB0xffff\fR
as \fIvlan\fR to match packets that are not tagged with a virtual LAN;
otherwise, specify a number between 0 and 4095, inclusive, as the
12-bit VLAN ID to match.

.IP \fBdl_src=\fImac\fR
Matches Ethernet source address \fImac\fR, which should be specified
as 6 pairs of hexadecimal digits delimited by colons,
e.g. \fB00:0A:E4:25:6B:B0\fR.

.IP \fBdl_dst=\fImac\fR
Matches Ethernet destination address \fImac\fR.

.IP \fBdl_type=\fIethertype\fR
Matches Ethernet protocol type \fIethertype\fR, which should be
specified as a integer between 0 and 65535, inclusive, either in
decimal or as a hexadecimal number prefixed by \fB0x\fR,
e.g. \fB0x0806\fR to match ARP packets.

.IP \fBnw_src=\fIip\fR[\fB/\fInetmask\fR]
Matches IPv4 source address \fIip\fR, which should be specified as an
IP address, e.g. \fB192.168.1.1\fR.  The optional \fInetmask\fR allows matching
only on an IPv4 address prefix.  The netmask is specificed "CIDR-style", i.e., 
\fB192.168.1.0/24\fR.

.IP \fBnw_dst=\fIip\fR[\fB/\fInetmask\fR]
Matches IPv4 destination address \fIip\fR.

.IP \fBnw_proto=\fIproto\fR
Matches IP protocol type \fIproto\fR, which should be specified as a
decimal number between 0 and 255, inclusive, e.g. 6 to match TCP
packets.

.IP \fBnw_tos=\fItos/dscp\fR
Matches ToS/DSCP (only 6-bits, not modify reserved 2-bits for future
use) field of IPv4 header \fItos/dscp\fR, which should be specified as
a decimal number between 0 and 255, inclusive.

.IP \fBtp_src=\fIport\fR
Matches transport-layer (e.g., TCP, UDP, ICMP) source port \fIport\fR,
which should be specified as a decimal number between 0 and 65535 (in
the case of TCP or UDP) or between 0 and 255 (in the case of ICMP),

RELEASE-NOTES
FlowVisor 0.7.2 : Mar 17 2011
    * added changeSlice fvctl/API command
    * fixed big bug in drop rules for disconnected controllers
    * fixed Null Pointer Exception race condition; stops crashing
    * added MRTG formatting scripts (thanks Chris Tengi!)
FlowVisor 0.7.1 : Feb 22 2011
    * diabled flow tracking by default
    * fixed bug in getSwitchFlowDB
FlowVisor 0.7 : 
    * fixed big I/O bug that caused switch flapping
    * added msg statistics, see `fvctl get{Slice,Switch}stats`
    * added flow tracking, see `fvctl get{Switch,SliceRewrite}FlowDB`
        to enable, run `fvctl setConfig flowvisor\!track_flows true`
    * lots of performance improvements
        * flow setup time ~5x improved
    * lots of small bugs fixed
    * password prompt no longer echos plaintext
    * better error handling in create slice and change flowspace
FlowVisor 0.6 - total rewrite (in Java!)
    * Single, unified flowspace config, not divided by slice
    * lldp is now rewritten with a trailer, not by rewritting the dst mac
    * OFError msgs now correctly encap the offending packet
    * NEW FlowVisor UserAPI: query and manipulate slice info


src/org/flowvisor/api/FVCtl.java
/**
 *
 */
package org.flowvisor.api;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.MalformedURLException;
import java.net.URL;
import java.security.KeyManagementException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.X509Certificate;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

import javax.net.ssl.HostnameVerifier;
import javax.net.ssl.HttpsURLConnection;
import javax.net.ssl.SSLContext;
import javax.net.ssl.SSLSession;
import javax.net.ssl.TrustManager;
import javax.net.ssl.X509TrustManager;

import org.apache.xmlrpc.XmlRpcException;
import org.apache.xmlrpc.client.XmlRpcClient;
import org.apache.xmlrpc.client.XmlRpcClientConfigImpl;
import org.flowvisor.api.FlowChange.FlowChangeOp;
import org.flowvisor.config.BracketParse;
import org.flowvisor.config.FVConfig;
import org.flowvisor.exceptions.MalformedFlowChange;
import org.flowvisor.exceptions.MapUnparsable;
import org.flowvisor.flows.FlowDBEntry;

/**
 * Client side stand alone command-line tool for invoking the FVUserAPI
 * 
 * This is pretty hacky and just for testing; people should write their own
 * clients and/or call the XMLRPC directly
 * 
 * @author capveg
 * 
 */
public class FVCtl {
	String URL;
	XmlRpcClientConfigImpl config;
	XmlRpcClient client;
	static APICmd[] cmdlist = new APICmd[] {
			new APICmd("listSlices", 0),
			new APICmd("createSlice", 3, "<slicename> <controller_url> <email>"),
			new APICmd("changeSlice", 3, "<slicename> <key> <value>"),
			new APICmd("deleteSlice", 1, "<slicename>"),
			new APICmd("changePasswd", 1, "<slicename>"),
			new APICmd("getSliceInfo", 1, "<slicename>"),

			new APICmd("getSliceStats", 1, "<slicename>"),
			new APICmd("getSwitchStats", 1, "<dpid>"),
			new APICmd("getSwitchFlowDB", 1, "<dpid>"),
			new APICmd("getSliceRewriteDB", 2, "<slicename> <dpid>"),

			new APICmd("listFlowSpace", 0),
			new APICmd("removeFlowSpace", 1, "<id>"),
			new APICmd("addFlowSpace", 4, "<dpid> <priority> <match> <actions>"),
			new APICmd("changeFlowSpace", 5,
					"<id> <dpid> <priority> <match> <actions>"),

			new APICmd("listDevices", 0),
			new APICmd("getDeviceInfo", 1, "<dpid>"),
			new APICmd("getLinks", 0),

			new APICmd("ping", 1, "<msg>"),
			new APICmd("getConfig", 1, "<configEntry>"),
			new APICmd("setConfig", 2, "<configEntry> <value>"),

			new APICmd("registerCallback", 2, "<URL> <cookie>"),
			new APICmd("unregisterCallback", 0), };

	static class APICmd {
		String name;
		int argCount;
		String usage;
		static HashMap<String, APICmd> cmdlist = new HashMap<String, APICmd>();

		APICmd(String name, int argCount, String usage) {
			this.name = name;
			this.argCount = argCount;
			this.usage = usage;
			cmdlist.put(name, this);
		}

		APICmd(String name, int argCount) {
			this(name, argCount, "");
		}

		@SuppressWarnings("unchecked")
		// Need to figure out magic java sauce to fix this
		void invoke(FVCtl client, String args[]) throws SecurityException,
				NoSuchMethodException, IllegalArgumentException,
				IllegalAccessException, InvocationTargetException {
			Class<String>[] params = new Class[args.length];
			for (int i = 0; i < args.length; i++)
				params[i] = String.class;
			Method m = FVCtl.class.getMethod("run_" + this.name, params);
			m.invoke(client, (Object[]) args);
		}
	}

	/**
	 * 
	 * @param URL
	 *            Server URL
	 */
	public FVCtl(String URL) {
		this.URL = URL;
	}

	/**
	 * Init connection to XMLRPC Server in URL
	 * 
	 * @throws MalformedURLException
	 * 
	 * @throws Exception
	 */
	public void init(String user, String passwd) throws MalformedURLException {
		this.installDumbTrust();
		config = new XmlRpcClientConfigImpl();
		config.setBasicUserName(user);
		config.setBasicPassword(passwd);
		config.setServerURL(new URL(this.URL));
		config.setEnabledForExtensions(true);

		client = new XmlRpcClient();
		// client.setTransportFactory(new
		// XmlRpcCommonsTransportFactory(client));
		// client.setTransportFactory(new )
		client.setConfig(config);
	}

	public void installDumbTrust() {

		// Create a trust manager that does not validate certificate chains
		// System.err.println("WARN: blindly trusting server cert - FIXME");
		TrustManager[] trustAllCerts = new TrustManager[] { new X509TrustManager() {
			public X509Certificate[] getAcceptedIssuers() {
				return null;
			}

			public void checkClientTrusted(X509Certificate[] certs,
					String authType) {
				// Trust always
			}

			public void checkServerTrusted(X509Certificate[] certs,
					String authType) {
				// Trust always
			}
		} };
		try {
			// Install the all-trusting trust manager
			SSLContext sc = SSLContext.getInstance("SSL");
			// Create empty HostnameVerifier
			HostnameVerifier hv = new HostnameVerifier() {
				public boolean verify(String arg0, SSLSession arg1) {
					return true;
				}
			};

			sc.init(null, trustAllCerts, new java.security.SecureRandom());
			HttpsURLConnection
					.setDefaultSSLSocketFactory(sc.getSocketFactory());
			HttpsURLConnection.setDefaultHostnameVerifier(hv);
		} catch (KeyManagementException e) {
			e.printStackTrace();
			throw new RuntimeException(e);
		} catch (NoSuchAlgorithmException e) {
			e.printStackTrace();
			throw new RuntimeException(e);
		}

	}

	public void run_listDevices() throws XmlRpcException {
		Object[] reply = (Object[]) this.client.execute("api.listDevices",
				new Object[] {});
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		for (int i = 0; i < reply.length; i++) {
			String dpid = (String) reply[i];
			System.out.println("Device " + i + ": " + dpid);
		}
	}

	@SuppressWarnings("unchecked")
	public void run_getDeviceInfo(String dpidStr) throws XmlRpcException {
		Map<String, Object> reply = (Map<String, Object>) this.client.execute(
				"api.getDeviceInfo", new Object[] { dpidStr });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		for (String key : reply.keySet()) {
			System.out.println(key + "=" + reply.get(key));
		}
	}

	public void run_getConfig(String name) throws XmlRpcException {

		Object reply = this.client.execute("api.getConfig",
				new Object[] { name });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		Object objects[] = (Object[]) reply;
		if (objects.length == 1)
			System.out.println(name + " = " + (String) objects[0]);
		else
			for (int i = 0; i < objects.length; i++)
				System.out
						.println(name + " " + i + " = " + (String) objects[i]);
	}

	public void run_setConfig(String name, String value) throws XmlRpcException {
		Object reply = this.client.execute("api.setConfig", new Object[] {
				name, value });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (!(reply instanceof Boolean)) {
			System.err.println("Didn't get boolean reply?; got" + reply);
			System.exit(-1);
		}
		boolean success = ((Boolean) reply).booleanValue();
		if (success) {
			System.out.println("success");
			System.exit(0);
		} else {
			System.out.println("failure");
			System.exit(-1);
		}
	}

	@SuppressWarnings("unchecked")
	public void run_getLinks() throws XmlRpcException, MapUnparsable {
		Object[] reply = (Object[]) this.client.execute("api.getLinks",
				new Object[] {});
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		Map<String, String> map;
		for (int i = 0; i < reply.length; i++) {
			if (!(reply[i] instanceof Map<?, ?>)) {
				System.err.println("not a map: Skipping unparsed reply: "
						+ reply[i]);
			} else {
				map = (Map<String, String>) reply[i];
				LinkAdvertisement ad = LinkAdvertisement.fromMap(map);
				System.out.println("Link " + i + ": " + ad);
			}
		}
	}

	@SuppressWarnings("unchecked")
	public void run_getSwitchFlowDB(String dpidString) throws XmlRpcException,
			MapUnparsable {
		Object[] reply = (Object[]) this.client.execute("api.getSwitchFlowDB",
				new Object[] { dpidString });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		Map<String, String> map;
		FlowDBEntry flowDBEntry;
		for (int i = 0; i < reply.length; i++) {
			if (!(reply[i] instanceof Map<?, ?>)) {
				System.err.println("not a map: Skipping unparsed reply: "
						+ reply[i]);
			} else {
				map = (Map<String, String>) reply[i];
				flowDBEntry = new FlowDBEntry();
				flowDBEntry.fromBacketMap(map);
				System.out.println("DBEntry " + i + ": " + flowDBEntry);
			}
		}
	}

	@SuppressWarnings("unchecked")
	public void run_getSliceRewriteDB(String sliceName, String dpidStr)
			throws XmlRpcException {

		Object ret = this.client.execute("api.getSliceRewriteDB", new Object[] {
				sliceName, dpidStr });
		Map<String, Object[]> flowRewriteDB;
		if (!(ret instanceof Map)) {
			throw new XmlRpcException("unknown reply type "
					+ ret.getClass().toString());
		}
		flowRewriteDB = (Map<String, Object[]>) ret;
		for (String original : flowRewriteDB.keySet()) {
			System.out.println("============ Original");
			System.out.println(original);
			System.out.println("\n=========== Rewritten to:");
			Object[] objs = flowRewriteDB.get(original);
			Map<String, String> rewrite;
			for (int i = 0; i < objs.length; i++) {
				rewrite = (Map<String, String>) objs[i];
				System.out.println("\t\t" + BracketParse.encode(rewrite));
			}

		}
	}

	public void run_changePasswd(String sliceName) throws IOException,
			XmlRpcException {
		String passwd = FVConfig.readPasswd("New password: ");
		Boolean reply = (Boolean) this.client.execute("api.changePasswd",
				new Object[] { sliceName, passwd });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply)
			System.out.println("success!");
		else
			System.err.println("failed!");
	}

	public void run_changeSlice(String sliceName, String key, String value)
			throws IOException, XmlRpcException {
		Boolean reply = (Boolean) this.client.execute("api.changeSlice",
				new Object[] { sliceName, key, value });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply)
			System.out.println("success!");
		else
			System.err.println("failed!");
	}

	@SuppressWarnings("unchecked")
	public void run_getSliceInfo(String sliceName) throws IOException,
			XmlRpcException {

		Object o = this.client.execute("api.getSliceInfo",
				new Object[] { sliceName });
		if (o == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		Map<String, String> reply = null;
		if (o instanceof Map<?, ?>)
			reply = (Map<String, String>) o;

		System.out.println("Got reply:");
		for (String key : reply.keySet())
			System.out.println(key + "=" + reply.get(key));
	}

	public void run_getSliceStats(String sliceName) throws IOException,
			XmlRpcException {

		Object o = this.client.execute("api.getSliceStats",
				new Object[] { sliceName });
		if (o == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		String reply = null;
		if (o instanceof String)
			reply = (String) o;

		System.out.println("Got reply:");
		System.out.println(reply);
	}

	public void run_getSwitchStats(String dpid) throws IOException,
			XmlRpcException {

		Object o = this.client.execute("api.getSwitchStats",
				new Object[] { dpid });
		if (o == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		String reply = null;
		if (o instanceof String)
			reply = (String) o;

		System.out.println("Got reply:");
		System.out.println(reply);
	}

	public void run_createSlice(String sliceName, String controller_url,
			String slice_email) throws IOException, XmlRpcException {
		String passwd = FVConfig.readPasswd("New password: ");
		Boolean reply = (Boolean) this.client
				.execute("api.createSlice", new Object[] { sliceName, passwd,
						controller_url, slice_email });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply)
			System.out.println("success!");
		else
			System.err.println("failed!");
	}

	public void run_ping(String msg) throws XmlRpcException {
		String reply = (String) this.client.execute("api.ping",
				new Object[] { msg });
		if (reply != null) {
			System.out.println("Got reply:");
			System.out.println(reply);
		} else {
			System.err.println("Got 'null' for reply :-(");
		}
	}

	public void run_deleteSlice(String sliceName) throws XmlRpcException {
		Boolean reply = (Boolean) this.client.execute("api.deleteSlice",
				new Object[] { sliceName });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply)
			System.out.println("success!");
		else
			System.err.println("failed!");
	}

	public void run_removeFlowSpace(String indexStr) throws XmlRpcException {
		FlowChange change = new FlowChange(FlowChangeOp.REMOVE,
				Integer.valueOf(indexStr));
		List<Map<String, String>> mapList = new LinkedList<Map<String, String>>();
		mapList.add(change.toMap());

		try {
			Object[] reply = (Object[]) this.client.execute(
					"api.changeFlowSpace", new Object[] { mapList });

			if (reply == null) {
				System.err.println("Got 'null' for reply :-(");
				System.exit(-1);
			}
			if (reply.length > 0)
				System.out.println("success: " + (String) reply[0]);
			else
				System.err.println("failed!");
		} catch (XmlRpcException e) {
			System.err.println("Failed: Flow Entry not found");
			System.exit(-1);
		}

	}

	public void run_addFlowSpace(String dpid, String priority, String match,
			String actions) throws XmlRpcException, MalformedFlowChange {
		do_flowSpaceChange(FlowChangeOp.ADD, dpid, null, priority, match,
				actions);
	}

	public void run_changeFlowSpace(String idStr, String dpid, String priority,
			String match, String actions) throws XmlRpcException,
			MalformedFlowChange {
		do_flowSpaceChange(FlowChangeOp.CHANGE, dpid, idStr, priority, match,
				actions);
	}

	private void do_flowSpaceChange(FlowChangeOp op, String dpid, String idStr,
			String priority, String match, String actions)
			throws XmlRpcException {
		if (match.equals("") || match.equals("any") || match.equals("all"))
			match = "OFMatch[]";
		Map<String, String> map = FlowChange.makeMap(op, dpid, idStr, priority,
				match, actions);

		try {
			FlowChange.fromMap(map);
		} catch (MalformedFlowChange e) {
			System.err.println("Local sanity check failed: " + e);
			return;
		}
		List<Map<String, String>> mapList = new LinkedList<Map<String, String>>();
		mapList.add(map);
		Object[] reply = (Object[]) this.client.execute("api.changeFlowSpace",
				new Object[] { mapList });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply.length > 0)
			System.out.println("success: " + (String) reply[0]);
		else
			System.err.println("failed!");
	}

	public void run_listSlices() throws XmlRpcException {
		Object[] reply = (Object[]) this.client.execute("api.listSlices",
				new Object[] {});
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		for (int i = 0; i < reply.length; i++) {
			String slice = (String) reply[i];
			System.out.println("Slice " + i + ": " + slice);
		}

	}

	public void run_listFlowSpace() throws XmlRpcException {
		Object[] result2 = (Object[]) client.execute("api.listFlowSpace",
				new Object[] {});
		if (result2 != null) {
			System.out.println("Got reply:");
			int i;
			for (i = 0; i < result2.length; i++)
				System.out.println("rule " + i + ": " + (String) result2[i]);
		} else {
			System.err.println("Got 'null' for reply :-(");
		}
	}

	public void run_registerCallback(String URL, String cookie)
			throws IOException, XmlRpcException, MalformedURLException {
		Boolean reply = (Boolean) this.client.execute(
				"api.registerTopologyChangeCallback", new Object[] { URL,
						cookie });
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply)
			System.out.println("success!");
		else
			System.err.println("failed!");
	}

	public void run_unregisterCallback() throws IOException, XmlRpcException,
			MalformedURLException {
		Boolean reply = (Boolean) this.client.execute(
				"api.unregisterTopologyChangeCallback", new Object[] {});
		if (reply == null) {
			System.err.println("Got 'null' for reply :-(");
			System.exit(-1);
		}
		if (reply)
			System.out.println("success!");
		else
			System.err.println("failed!");
	}

	private static void usage(String string) {
		usage(string, true);
	}

	private static void usage(String string, boolean printFull) {
		System.err.println(string);
		if (printFull) {
			System.err
					.println("Usage: FVCtl [--debug=true] [--user=user] [--url=url] "
							+ "[--passwd-file=filename] command [args...] ");
			for (int i = 0; i < FVCtl.cmdlist.length; i++) {
				APICmd cmd = FVCtl.cmdlist[i];
				System.err.println("\t" + cmd.name + " " + cmd.usage);
			}
		}
		System.exit(-1);
	}

	/**
	 * Front-end cmdline parser for FVCtl
	 * 
	 * @param args
	 * @throws SecurityException
	 * @throws IllegalArgumentException
	 * @throws NoSuchMethodException
	 * @throws IllegalAccessException
	 * @throws InvocationTargetException
	 * @throws IOException
	 */
	public static void main(String args[]) {
		// FIXME: make URL a parameter
		// FVCtl client = new
		// FVCtl("https://root:joemama@localhost:8080/xmlrpc");
		String URL = "https://localhost:8080/xmlrpc";
		String user = "root";
		String passwd = null;
		boolean debug = false;

		int cmdIndex = 0;
		// FIXME: find a decent java cmdline args parsing lib
		while ((args.length > cmdIndex) && (args[cmdIndex].startsWith("--"))) {
			String params[] = args[cmdIndex].split("=");
			if (params.length < 2)
				usage("parameter " + params[0] + " needs an argument");
			if (params[0].equals("--url"))
				URL = params[1];
			else if (params[0].equals("--user"))
				user = params[1];
			else if (params[0].equals("--debug"))
				debug = Boolean.valueOf(params[1]);
			else if (params[0].equals("--passwd-file")) {

				try {
					passwd = new BufferedReader(new FileReader(new File(
							params[1]))).readLine();
				} catch (FileNotFoundException e) {
					die(debug, "file: '" + params[1] + "' :: ", e);
				} catch (IOException e) {
					die(debug, "IO: ", e);
				}
			} else
				usage("unknown parameter: " + params[0]);
			cmdIndex++;
		}
		if (args.length == cmdIndex)
			usage("need to specify a command");

		APICmd cmd = APICmd.cmdlist.get(args[cmdIndex]);
		if (cmd == null)
			usage("command '" + args[cmdIndex] + "' does not exist");
		if ((args.length - 1 - cmdIndex) < cmd.argCount)
			usage("command '" + args[cmdIndex] + "' takes " + cmd.argCount
					+ " args: only " + (args.length - 1 - cmdIndex)
					+ " given\n" + args[cmdIndex] + " " + cmd.usage, false);
		String[] strippedArgs = new String[args.length - 1 - cmdIndex];
		System.arraycopy(args, cmdIndex + 1, strippedArgs, 0,
				strippedArgs.length);
		try {
			if (passwd == null)
				passwd = FVConfig.readPasswd("Enter " + user + "'s passwd: ");
			FVCtl client = new FVCtl(URL);
			client.init(user, passwd);
			cmd.invoke(client, strippedArgs);
		} catch (Exception e) {
			die(debug, "error: ", e);
		}
	}

	private static void die(boolean debug, String string, Exception e) {
		Throwable cause = e;
		while (cause.getCause() != null) {
			if (debug)
				cause.printStackTrace(System.err);
			cause = cause.getCause();
		}
		if (debug)
			cause.printStackTrace(System.err);
		System.err.println(string + cause);
		System.exit(-1);
	}

}
